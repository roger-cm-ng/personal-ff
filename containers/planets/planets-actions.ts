import { Dispatch } from 'redux';
import { getPlanets } from '../../helpers/resources';
import { inject } from './planets-utils';
import { ERROR_RAISED } from '../error-handler/error-handler-actions';
import ErrorRestart from '../../components/error-content/error-restart/error-restart';

export const PLANETS_FETCHED = 'PLANETS_FETCHED';
export const PLANET_CHOSEN = 'PLANET_CHOSEN';

export const fetchPlanets = () => async (dispatch: Dispatch) => {
  try {
    const payload = await getPlanets();

    const { data, status } = payload;

    if (status >= 200 && status < 300) {
      dispatch({
        type: PLANETS_FETCHED,
        payload: inject(data)
      });
    } else {
      dispatch({
        type: ERROR_RAISED,
        component: ErrorRestart
      });
    }
  } catch (err) {
    dispatch({
      type: ERROR_RAISED,
      component: ErrorRestart
    });
  }
};

export const choosePlanet = (planet: string[]) => ({
  type: PLANET_CHOSEN,
  planet
});
