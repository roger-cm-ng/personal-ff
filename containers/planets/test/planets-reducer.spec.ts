import planetsReducer from '../planets-reducer';
import { PLANETS_FETCHED, PLANET_CHOSEN } from './../planets-actions';
import {
  PLANETS_MOCK,
  PLANET_EMPTY_DONLON_MOCK,
  PLANET_EMPTY_ENCHAI_MOCK,
  PLANET_DONLON_EMPTY_MOCK,
  PLANET_ENCHAI_LERBIN_MOCK
} from './planets-mock-data';

describe('planets reducer', () => {
  const initialState = () => planetsReducer([], {
    type: PLANETS_FETCHED,
    payload: []
  });

  const acquiredState = () => planetsReducer(initialState(), {
    type: PLANETS_FETCHED,
    payload: PLANETS_MOCK
  });

  const emptyDonlonState = () => planetsReducer(acquiredState(), {
    type: PLANET_CHOSEN,
    planet: ['', 'Donlon']
  });

  const emptyEnchaiState = () => planetsReducer(emptyDonlonState(), {
    type: PLANET_CHOSEN,
    planet: ['', 'Enchai']
  });

  const donlonEmptyState = () => planetsReducer(emptyEnchaiState(), {
    type: PLANET_CHOSEN,
    planet: ['Donlon', '']
  });

  const enchaiLerbinState = () => planetsReducer(donlonEmptyState(), {
    type: PLANET_CHOSEN,
    planet: ['Enchai', 'Lerbin']
  });

  it('should return initial state', () => {
    expect(initialState()).toHaveLength(0);
  });

  it('should update the state with acquired data when type is PLANETS_FETCHED', () => {
    expect(acquiredState()).toMatchSnapshot();
    expect(acquiredState()).toEqual(PLANETS_MOCK);
  });

  it('should update the state with planet from empty to Donlon', () => {
    expect(emptyDonlonState()).toEqual(PLANET_EMPTY_DONLON_MOCK);
  });

  it('should update the state with planet from empty to Enchai', () => {
    expect(emptyEnchaiState()).toEqual(PLANET_EMPTY_ENCHAI_MOCK);
  });

  it('should update the state with planet from Donlon to empty', () => {
    expect(donlonEmptyState()).toEqual(PLANET_DONLON_EMPTY_MOCK);
  });

  it('should update the state with planet from Enchai to Lerbin', () => {
    expect(enchaiLerbinState()).toEqual(PLANET_ENCHAI_LERBIN_MOCK);
  });
});
