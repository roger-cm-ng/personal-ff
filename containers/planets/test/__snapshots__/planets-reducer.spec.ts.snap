// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`planets reducer should update the state with acquired data when type is PLANETS_FETCHED 1`] = `
Array [
  Object {
    "distance": 100,
    "isChosen": false,
    "name": "Donlon",
  },
  Object {
    "distance": 200,
    "isChosen": false,
    "name": "Enchai",
  },
  Object {
    "distance": 300,
    "isChosen": false,
    "name": "Jebing",
  },
  Object {
    "distance": 400,
    "isChosen": false,
    "name": "Sapir",
  },
  Object {
    "distance": 500,
    "isChosen": false,
    "name": "Lerbin",
  },
  Object {
    "distance": 600,
    "isChosen": false,
    "name": "Pingasor",
  },
]
`;
